integration/container_stats_test.go

```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
+)
+
+// Test to verify for a container ID
+func TestContainerStats(t *testing.T) {
+	t.Logf("Create a pod config and run sandbox container")
+	sbConfig := PodSandboxConfig("sandbox1", "stats")
+	sb, err := runtimeService.RunPodSandbox(sbConfig)
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, runtimeService.StopPodSandbox(sb))
+		assert.NoError(t, runtimeService.RemovePodSandbox(sb))
+	}()
+	t.Logf("Create a container config and run container in a pod")
+	containerConfig := ContainerConfig(
+		"container1",
+		pauseImage,
+		WithTestLabels(),
+		WithTestAnnotations(),
+	)
+	cn, err := runtimeService.CreateContainer(sb, containerConfig, sbConfig)
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, runtimeService.RemoveContainer(cn))
+	}()
+	require.NoError(t, runtimeService.StartContainer(cn))
+	defer func() {
+		assert.NoError(t, runtimeService.StopContainer(cn, 10))
+	}()
+	t.Logf("Fetch stats for container")
+	var s *runtime.ContainerStats
+	require.NoError(t, Eventually(func() (bool, error) {
+		s, err = runtimeService.ContainerStats(cn)
+		if err != nil {
+			return false, err
+		}
+		if s.GetWritableLayer().GetUsedBytes().GetValue() != 0 &&
+			s.GetWritableLayer().GetInodesUsed().GetValue() != 0 {
+			return true, nil
+		}
+		return false, nil
+	}, time.Second, 30*time.Second))
+
+	t.Logf("Verify stats received for container %q", cn)
+	testStats(t, s, containerConfig)
+}
+
+// Test to verify filtering without any filter
+func TestContainerListStats(t *testing.T) {
+	t.Logf("Create a pod config and run sandbox container")
+	sbConfig := PodSandboxConfig("running-pod", "statsls")
+	sb, err := runtimeService.RunPodSandbox(sbConfig)
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, runtimeService.StopPodSandbox(sb))
+		assert.NoError(t, runtimeService.RemovePodSandbox(sb))
+	}()
+	t.Logf("Create a container config and run containers in a pod")
+	containerConfigMap := make(map[string]*runtime.ContainerConfig)
+	for i := 0; i < 3; i++ {
```

Comment:
We should have a dead container in the test, and for dead container, `testStats` won't work.
Probably add a TODO here, and you may want to manually validate it first.

Reply:
had added on top of testStats coz that needed change. I have manually verified the case for dead containers by introducing a stopContainer and checking the stats.
```
=== RUN   TestContainerListStatsDead
&ContainerStats{Attributes:&ContainerAttributes{Id:6bf3066a1929839b5a5802cf3b3b49c7c292cc53364b3280d2a0a07dc24d9f77,Metadata:&ContainerMetadata{Name:container0,Attempt:0,},Labels:map[string]string{key: value,},Annotations:map[string]string{a.b.c: test,},},Cpu:nil,Memory:nil,WritableLayer:&FilesystemUsage{Timestamp:1506452046974590603,StorageId:&StorageIdentifier{Uuid:c656051a-9391-40fe-8333-57b410884232,},UsedBytes:&UInt64Value{Value:7,},InodesUsed:&UInt64Value{Value:20480,},},}
&ContainerStats{Attributes:&ContainerAttributes{Id:178a2722bd4ecec8728a952f07474b7714896f4afe8d69b3b58989b387e45804,Metadata:&ContainerMetadata{Name:container1,Attempt:0,},Labels:map[string]string{key: value,},Annotations:map[string]string{a.b.c: test,},},Cpu:nil,Memory:nil,WritableLayer:&FilesystemUsage{Timestamp:1506452046972987825,StorageId:&StorageIdentifier{Uuid:c656051a-9391-40fe-8333-57b410884232,},UsedBytes:&UInt64Value{Value:7,},InodesUsed:&UInt64Value{Value:20480,},},}
&ContainerStats{Attributes:&ContainerAttributes{Id:52e891cd436a10d71bc37d71f978d4142c4d49b7922ca7a2442ca75fe15c10b8,Metadata:&ContainerMetadata{Name:container2,Attempt:0,},Labels:map[string]string{key: value,},Annotations:map[string]string{a.b.c: test,},},Cpu:nil,Memory:nil,WritableLayer:&FilesystemUsage{Timestamp:1506452046973408844,StorageId:&StorageIdentifier{Uuid:c656051a-9391-40fe-8333-57b410884232,},UsedBytes:&UInt64Value{Value:7,},InodesUsed:&UInt64Value{Value:20480,},},}
--- PASS: TestContainerListStatsDead (32.88s)
```

Reply:
Thanks!